packages/design-system/src/components/InlineMessage/InlineMessage.tsx

```diff
@@ -1,67 +1,58 @@
 
-import * as S from './InlineMessage.style';
-import { light } from '../../themes';
+import { InlineMessageVariantType } from './Primitive/InlineMessagePrimitive';
 
-export type InlineMessageProps = PropsWithChildren<any> &
-	StyledProps<any> & {
-		/** The icon element to display */
-		icon: IconName;
-		/** The title of the message */
-		title?: string;
-		/** The content of the message */
-		description: string;
-		/** Size variant */
-		small: boolean;
-		/** Link element at the end */
-		link?: React.ReactElement;
-		/** If it is an inline message or a block with a background */
-		withBackground?: boolean;
-	};
+type Information = InlineMessageVariantType<'information', InlineMessageInformationProps>;
+type Success = InlineMessageVariantType<'success', InlineMessageSuccessProps>;
+type Warning = InlineMessageVariantType<'warning', InlineMessageWarningProps>;
+type Destructive = InlineMessageVariantType<'destructive', InlineMessageDestructiveProps>;
+type Beta = InlineMessageVariantType<'beta', InlineMessageBetaProps>;
 
-/**
- Inline message highlights information necessary to display for the user in many different contexts.
- It can be additional information related to system status, it can be a required action to complete the current task.
- @link https://inclusive-components.design/notifications
- * */
-const InlineMessage = React.forwardRef(
-	(
-		{ icon, title, description, link, small, withBackground, theme, ...rest }: InlineMessageProps,
-		ref,
-	) => (
-		<S.InlineMessage
-			role="status"
-			aria-live="polite"
-			{...rest}
-			small={small}
-			withBackground={withBackground}
-			theme={withBackground ? light : theme}
-			ref={ref}
-		>
-			{icon && (
-				<S.Icon small={small}>
-					<Icon name={icon} />
-				</S.Icon>
-			)}
-			<S.Content>
-				{title && (
-					<>
-						<S.Title>{title}</S.Title>{' '}
-					</>
-				)}
-				<>
-					<S.Description>{description}</S.Description>{' '}
-				</>
-				{link && (
-					<S.Link>{React.cloneElement(link, { theme: withBackground ? light : theme })}</S.Link>
-				)}
-			</S.Content>
-		</S.InlineMessage>
-	),
-);
+export type InlineMessageType = Information | Success | Warning | Destructive | Beta;
+
+const InlineMessage = forwardRef((props: InlineMessageType, ref: Ref<HTMLDivElement>) => {
+	switch (props.variant) {
+		case 'information': {
+			const { variant, ...rest } = props;
+			return <InlineMessageInformation {...rest} ref={ref} />;
+		}
+
+		case 'destructive': {
+			const { variant, ...rest } = props;
+			return <InlineMessageDestructive {...rest} ref={ref} />;
+		}
+
+		case 'success': {
+			const { variant, ...rest } = props;
+			return <InlineMessageSuccess {...rest} ref={ref} />;
+		}
+
+		case 'warning': {
+			const { variant, ...rest } = props;
+			return <InlineMessageWarning {...rest} ref={ref} />;
+		}
+
+		case 'beta': {
+			const { variant, ...rest } = props;
+			return <InlineMessageBeta {...rest} ref={ref} />;
+		}
```

Comment:
```suggestion
	const { variant, ...rest } = props;

	switch (props.variant) {
		case 'information':
			return <InlineMessageInformation {...rest} ref={ref} />;

		case 'destructive': 
			return <InlineMessageDestructive {...rest} ref={ref} />;

		case 'success': {
			const { variant, ...rest } = props;
			return <InlineMessageSuccess {...rest} ref={ref} />;
		}

		case 'warning': 
			return <InlineMessageWarning {...rest} ref={ref} />;

		case 'beta':
			return <InlineMessageBeta {...rest} ref={ref} />;
```

Reply:
Good instincts but this doesn't work with TS. 
If you destructure outside of the `case`, TS cannot infer the type of `...rest` anymore.